SparseLinear.lua

```diff
@@ -51,6 +51,9 @@ function SparseLinear:reshapeInput(input)
 end
 
 function SparseLinear:updateOutput(input)
+   if self.sparseUpdate == ONE_LAST_INPUT then
+      self.sparseUpdate = ACC_MULTIPLE_TIMES
```

Comment:
Yes that is the intended behavior. After enough non zero elements have been
passed through it becomes more efficient to simply update every parameter
instead of finding the unique nonzeroes first. A smarter method would
probably be to do this transition after accumulating too many non zero
elements, but to simplify it, my design decision was to only do sparse
updates after one minibatch.

On Sun, Mar 20, 2016, 14:01 Soumith Chintala <notifications@github.com>
wrote:

> In SparseLinear.lua
> <https://github.com/torch/nn/pull/725#discussion_r56768682>:
>
> > @@ -51,6 +51,9 @@ function SparseLinear:reshapeInput(input)
> >  end
> >
> >  function SparseLinear:updateOutput(input)
> > +   if self.sparseUpdate == ONE_LAST_INPUT then
> > +      self.sparseUpdate = ACC_MULTIPLE_TIMES
>
> this is wrong. afaik this has to be self.sparseUpdate = NO_LAST_INPUT
> The rewritten logic in this file actually never goes back to
> self.sparseUpdate being reset to self.sparseUpdate = NO_LAST_INPUT as lines
> 195 and 208 have been removed.
> Practically, this PR will enable sparse optimizations only for the first
> mini-batch, and take the dense path afterwards because it is always be in
> ACC_MULTIPLE_TIMES mode.
> However, I suspect that I misunderstand this PR, can you explain the state
> transitions and their expected behavior?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/torch/nn/pull/725/files/1156255899c621e322a1f7991a817d4139af754d#r56768682>
>